src/Lumper.Lib/BSP/BspFile.cs

```diff
@@ -35,45 +43,238 @@ public class BspFile
+        FilePath = GetUnescapedFilePathString(path);
+        FileStream = File.OpenRead(FilePath);
+
+        using var reader = new BspFileReader(this, FileStream, handler);
+
+        // UI/CLI should track whether this gets cancelled and dispose of this class, it's in a useless
+        // half-loaded state and mustn't be used.
+        if (!reader.Load())
+        {
+            Logger.Info("Loading cancelled");
+            return null;
+        }
+
+        Logger.Info($"Loaded {Name}.bsp");
+        return this;
     }
 
-    public void Load(Stream stream)
+    public BspFile? Load(IoHandler handler, Stream stream)
     {
-        FilePath = null;
-        Stream.Dispose();
-        Stream = new MemoryStream();
-        stream.CopyTo(Stream);
-        var reader = new BspFileReader(this, Stream);
-        reader.Load();
+        using var reader = new BspFileReader(this, stream, handler);
+
+        if (!reader.Load())
+        {
+            Logger.Info("Loading cancelled");
+            return null;
+        }
+
+        Logger.Info("Loaded BSP from stream");
+        return this;
     }
 
-    public void Save(string path)
+    public void SaveToFile(
+        IoHandler handler,
+        string? path,
+        DesiredCompression compress,
+        bool makeBackup = true)
     {
-        if (path == FilePath)
-            throw new IOException("Can't write BSP to the same file");
-        using FileStream stream = File.Open(path, FileMode.Create);
-        Save(stream);
+        // If you add something here, also add it to the BspReader
+        if (path is null && FilePath is null)
+            throw new ArgumentException("Not given a path to write to, and current BSP doesn't have a path");
+
+        string outPath;
+        string? backupPath = null;
+        string? escapedPath;
+        if (path is not null &&
+            (escapedPath = GetUnescapedFilePathString(path)) != FilePath)
+        {
+            outPath = escapedPath;
+        }
+        else
+        {
+            outPath = FilePath!;
+
+            // Path is null so probably overwriting a file. If so, make a backup.
+            // This program isn't well-tested, don't want it mangling someones's BSP.
+            // Only do if overwriting existing file i.e. string != null
+            if (makeBackup && File.Exists(outPath))
+            {
+                backupPath = BspExtensionRegex().Replace(outPath, "") +
+                             $"_lumperbackup{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}.bsp";
+                if (File.Exists(backupPath))
+                {
+                    // This should never happen since path depends on current
+                    // millisecond but whatever
+                    Logger.Error($"Backup path {backupPath} exists!");
+                    return;
+                }
+
+                try
+                {
+                    // TODO: go back to copy?
```

Comment:
File.Copy instead of stream stuff or what do you mean?

Reply:
Woops, silly little comment for myself I left in